### ISSUE TITLE
VMs created with virt-install --cloud-init can't be started again, because of a missing cloudinit.iso in the CDROM drive

### ISSUE THREAD

**Issue by kparal** [virt-manager/virt-manager#178]
The new `virt-install --cloud-init` option works well for spawning a new Fedora Cloud VM. However, once the VM is shut down, it can't be started again. The reason seems to be a missing `*-cloudinit.iso` image that is assigned to the CDROM drive, but the actual image has been deleted in the meantime.

Here is me creating a new VM (in `qemu:///system`, but the same problem exists even in `qemu:///session`):
```
$ ./virt-install --name cloud1 --memory 2000 --boot hd --disk=size=10,backing_store="/data/iso/final/Fedora-Cloud-Base-33-1.2.x86_64.qcow2" --cloud-init
WARNING  Defaulting to --cloud-init root-password-generate=yes,disable=yes
WARNING  No operating system detected, VM performance may suffer. Specify an OS with --os-variant for optimal results.

Starting install...
Password for first root login is: utMTYf96obBrFCaM
Installation will continue in 10 seconds (press Enter to skip)...
Allocating 'virtinst-erl3cbv1-cloudinit.iso'                                                   | 364 kB  00:00:00     
Transferring virtinst-erl3cbv1-cloudinit.iso                                                   | 364 kB  00:00:00     
Allocating 'cloud1.qcow2'                                                                      |  10 GB  00:00:00     
Running text console command: virsh --connect qemu:///system console cloud1
Connected to domain cloud1
Escape character is ^] (Ctrl + ])
... <snip> ...
$ poweroff
... <snip> ...
Domain creation completed.
You can restart your domain by running:
  virsh --connect qemu:///system start cloud1
```

And when I try to start it again:
```
$ virsh start cloud1
error: Failed to start domain cloud1
error: Cannot access storage file '/var/lib/libvirt/boot/virtinst-erl3cbv1-cloudinit.iso': No such file or directory
```

The same error is in virt-manager:
![cloud-init](https://user-images.githubusercontent.com/755451/98238327-84cbb500-1f66-11eb-8aa1-6a197df63cdf.png)

In virt-manager, I see that `IDE CDROM 1` has a `Source path` set to `/var/lib/libvirt/boot/virtinst-erl3cbv1-cloudinit.iso`, but the file doesn't exist.

It seems to be that either:
a) the `cloudinit.iso` should not get deleted after the first boot, but should be kept on disk, and can be deleted when the actual VM gets deleted (the same how it works for a hard drive); or
b) the `cloudinit.iso` should get deleted, but also the CDROM path should be reset to an empty value at the same time, so that future boots are not broken

I tested clearing the CDROM path and my VM boots fine, which suggests that b) could be the right solution.

**Distro**: Fedora 33
Happens with both distributions package `virt-install-3.1.0-1.fc33.noarch` and with the latest git checkout `04ad8df752c11ef5`.

**Comment by crobinso** [comment]
Thanks for the report. Can you provide full `virt-install --debug` output when reproducing? In a simple test I couldn't reproduce

**Comment by kparal** [comment]
That's interesting, I see it every time and my colleague as well.

Here's the install log from `~/.cache/virt-manager/virt-install.log`:
[virt-install.log](https://github.com/virt-manager/virt-manager/files/5509976/virt-install.log)

Here's the generated cloud1 VM xml from `virsh dumpxml cloud1`:
[cloud1.zip](https://github.com/virt-manager/virt-manager/files/5509985/cloud1.zip)

In the XML I see:
```
    <disk type='file' device='cdrom'>
      <driver name='qemu' type='raw'/>
      <source file='/var/lib/libvirt/boot/virtinst-w2tecqgk-cloudinit.iso'/>
      <target dev='hdb' bus='ide'/>
      <readonly/>
      <address type='drive' controller='0' bus='0' target='0' unit='1'/>
    </disk>
```
The file is not present:
```
$ ls /var/lib/libvirt/boot/virtinst-w2tecqgk-cloudinit.iso
ls: cannot access '/var/lib/libvirt/boot/virtinst-w2tecqgk-cloudinit.iso': No such file or directory
```
In the log I see:
```
[Mon, 09 Nov 2020 12:03:41 virt-install 13409] DEBUG (installer:400) Removing volume 'virtinst-w2tecqgk-cloudinit.iso'
[Mon, 09 Nov 2020 12:03:41 virt-install 13409] DEBUG (installer:405) Removing /home/kparal/.cache/virt-manager/boot/virtinst-txac84r3-meta-data
[Mon, 09 Nov 2020 12:03:41 virt-install 13409] DEBUG (installer:405) Removing /home/kparal/.cache/virt-manager/boot/virtinst-ff207ucy-user-data
[Mon, 09 Nov 2020 12:03:41 virt-install 13409] DEBUG (installer:405) Removing /home/kparal/.cache/virt-manager/boot/virtinst-w2tecqgk-cloudinit.iso
```

Huh, that's interesting, the paths are different - `/var/lib/libvirt/boot/` vs `/home/kparal/.cache/virt-manager/boot/`. The second location is empty as well:
```
$ ls ~/.cache/virt-manager/boot/
$
```

Perhaps the difference in location wouldn't even matter if the CDROM got unassigned after shutdown.

**Comment by crobinso** [comment]
I can reproduce, my test was busted, this only triggers with certain install option code paths in virt-install. I'll fix it shortly

**Comment by emmceemoore** [comment]
I was bitten by a variation of this issue on `4.1.0`. (With `package_reboot_if_required: true` in my `--cloud-init` "user-data" I ended up with a similar error.)

Given that cloud-init runs on _more_ than just he first boot, I wonder if it would make sense to keep the generated ISO and CDROM device around instead of trying to figure out the semantic "first boot" and clean things up afterwards. Just a thought.

**Comment by TomCab** [comment]
I found the same issue on `4.1.0` running on Debian 12 (Bookworm) with no backports. This is an old thread for a problem that has been solved in `virt-install` versions > `4.1.0`,  but finding a work-around cost me a bit of time, since installing backports is not an option in my case. Maybe someone else will find this work-around useful.

Running `virt-install` version `4.1.0`  with `--debug` shows that `--cloud-init user-data=/path/to/file` triggers copying `user-data` into an `.iso` temporary image, which is then loaded at first virtual machine boot; this is the expected behavior. At the end of the VM installation however, the temporary `.iso` image is removed from `~/.cache/virt-manager/boot/` and with it goes all modified login information included in `user-data` and login is impossible.

Work-around:

1. Create one file called `user-data` and another called `meta-data`  containing the changes you want to inject into the virtual machine.
2. The file `meta-data` can be empty  if you do not want to include any metadata, but it is necessary for this work-around to be successful.
3. Create a `<data_injection>.iso` image containing only those two files in its file system's root:

```shell
xorriso -as genisoimage -output <data_injection>.iso -volid CIDATA -joliet -rock user-data meta-data
```

4. Pass the `<data_injection>.iso` image thus created as option to `virt-install` and specify that it is to be mounted as a CD-ROM during VM installation:

```shell
virt-install \
    --connect qemu:///system \
    --name <vm_name> \
    --memory <vm_memory> \
    --machine <vm_machine> \
    --arch <vm_arch> \
    --vcpus <vm_vcpus> \
    --cpu <vm_cpu> \
    --import \
    --osinfo name=<virt_os_name> \
    --disk <distro_image_disk> \
    --disk <data_injection>.iso,device=cdrom \
    --virt-type qemu \
    --hvm
```

The VM thus created successfully imports the modifications included in files  `user-data` and `meta-data`.
